lib/iterator.js

```diff
@@ -0,0 +1,312 @@
+  constructor(base, predicate, thisArg) {
+    super(base);
+    this.predicate = predicate;
+    this.thisArg = thisArg;
+  }
+
+  next() {
+    for (const value of this.base) {
+      if (this.predicate.call(this.thisArg, value)) {
+        return { done: false, value };
+      }
+    }
+    return { done: true, value: undefined };
+  }
+}
+
+class FlatIterator extends Iterator {
+  constructor(base, depth) {
+    super(base);
+    this.depth = depth;
+    this.stack = [base];
+  }
+
+  next() {
+    while (this.stack.length) {
+      const top = this.stack[this.stack.length - 1];
+      const next = top.next();
+
+      if (next.done) {
+        this.stack.pop();
+        continue;
+      }
+
+      if (this.depth === this.stack.length - 1) {
+        return next;
+      }
+
+      if (next.value[Symbol.iterator]) {
+        this.stack.push(next.value[Symbol.iterator]());
+        continue;
+      }
+
+      if (!next.value.next) {
+        return next;
+      }
+
+      this.stack.push(next);
+    }
+
+    return { done: true, value: undefined };
+  }
+}
+
+class FlatMapIterator extends Iterator {
+  constructor(base, mapper, thisArg) {
+    super(base);
+    this.mapper = mapper;
+    this.thisArg = thisArg;
+    this.currentIterator = undefined;
+  }
+
+  next() {
+    if (!this.currentIterator) {
+      const next = this.base.next();
+      if (next.done) {
+        return next;
+      }
+      this.currentIterator = new Iterator(
+        this.mapper.call(this.thisArg, next.value)
+      );
+    }
+
+    const next = this.currentIterator.next();
+
+    if (next.done) {
+      this.currentIterator = undefined;
+      return this.next();
+    }
+    return next;
+  }
+}
+
+class TakeIterator extends Iterator {
+  constructor(base, amount) {
+    super(base);
+    this.amount = amount;
+    this.iterated = 0;
+  }
+
+  next() {
+    if (this.iterated++ < this.amount) {
+      return this.base.next();
+    }
+    return { done: true, value: undefined };
+  }
+}
+
+class ZipIterator extends Iterator {
+  constructor(base, iterators) {
+    super(base);
+    this.iterators = iterators.map(toIterator);
+  }
+
+  next() {
+    const result = [];
+
+    const next = this.base.next();
+    if (next.done) {
+      return next;
+    }
+    result.push(next.value);
+
+    for (const iterator of this.iterators) {
+      const next = iterator.next();
+      if (next.done) {
+        return next;
+      }
+      result.push(next.value);
+    }
+    return { done: false, value: result };
+  }
+}
+
+class JoinIterator extends Iterator {
+  constructor(base, iterators) {
+    super(base);
+    this.currentIterator = base;
+    this.iterators = iterators.map(toIterator);
+  }
+
+  next() {
+    const next = this.currentIterator.next();
+    if (!next.done) {
+      return next;
+    }
+    const { done, value: iterator } = this.iterators.next();
+    if (done) {
+      return next;
+    }
+    this.currentIterator = iterator;
+    return this.next();
```

Comment:
Can you do this without recursion?
This seems to lead to getting an error with maximum call stack size exceedance.

Reply:
This looks like premature optimisations. You can change that in a follow-up PR. You'll need to join a lot of empty iterators